platforms/software/testing-base-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java

```diff
@@ -137,13 +178,42 @@ private boolean matchesClassAndMethod(List<TestPattern> patterns, String classNa
         return false;
     }
 
-    private static class TestPattern {
-        private Pattern pattern;
+    @NullMarked
+    private static final class FileTestPattern {
+        private final Pattern pattern;
+
+        private FileTestPattern(String path) {
+            pattern = preparePattern(path);
+        }
+
+        private boolean matches(String input) {
+            return pattern.matcher(input).matches();
+        }
+
+        private static Pattern preparePattern(String input) {
+            StringBuilder pattern = new StringBuilder();
+            String[] split = StringUtils.splitPreserveAllTokens(input, '*');
+            for (String s : split) {
+                if (s.isEmpty()) {
+                    pattern.append(".*"); //replace wildcard '*' with '.*'
+                } else {
+                    if (pattern.length() > 0) {
+                        pattern.append(".*"); //replace wildcard '*' with '.*'
+                    }
+                    pattern.append(Pattern.quote(s)); //quote everything else
+                }
+            }
+            return Pattern.compile(pattern.toString());
+        }
+    }
```

Comment:
I _think_ this is the only bit of tweaking we need to turn the filter pattern into a useful regex.